### PR TITLE
Issue #9194: Client authentication fails when using signed JWT, if the JWA signing algorithm is not RS256

### PR DESCRIPTION
**Pull request by rmartinc**
Issue: #9194 

The PR fixes the `ClientPublicKeyLoader` to not set RSA and RS256 as the compulsory type and algorithm in the `KeyWrapper`. Related changes:

* Client asymmetric and ECDSA classes now checks the type of key algorithm (RSA and EC respectively) and assign the crypto alg if not specified by the certificate loader (previously only the first class did that).
* The `InfinispanPublicKeyStorageProvider` returns a clone of the key because if not the alg is set by the previous point and no other algorithm is allowed for the client while cached. This way, if the definition of the client certificate does not enforce it, the client can use any algorithm allowed by the key type (RSA or EC).
* The `JWTClientCredentialsProvider` has been slightly modified to pass the _algorithm_ option for the credentials and not force RS256. This was done for the tests but is now allowed for any java client too. If you want that I modify the documentation (this [chapter](https://github.com/keycloak/keycloak-documentation/blob/main/securing_apps/topics/oidc/java/client-authentication.adoc)), please let me know.
* Two tests modified to cover the changes.

@mposolda Please take a look to this when you have time.

Thanks!

### CODE REVIEW

**Comment by rmartinc** [comment]
@mposolda Please take a look to this PR when you have time or redirect it to other reviewer. Thanks a lot!

**Comment by arthurchan35** [comment]
Hi @rmartinc @mposolda 

Is there any update for this PR?

**Comment by rmartinc** [comment]
Hi! @stianst @mposolda is there a chance to assign this PR to someone for review? I can rebase it if needed. Thanks a lot!

**Comment by rmartinc** [comment]
@mposolda @stianst Are you interested in this PR? I was going to rebase and recheck it because it is quite old now. :smiley:

**Comment by mposolda** [comment]
@tnorimat Do you please have a chance to review this PR since you are very familiar with this part of the codebase?

**Comment by tnorimat** [comment]
@mposolda Yes, I will review this PR.

**Comment by mposolda** [comment]
@tnorimat Thanks!

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/ClientAuthSignedJWTTest.java [hunk 1]
```diff
@@ -494,14 +524,14 @@ public void testSuccessDefaultAlgWhenNoAlgSetInJWK() throws Exception {
             OAuthClient.AccessTokenResponse response = doGrantAccessTokenRequest("test-user@localhost", "password", createSignedRequestToken("client2", getRealmInfoUrl(), privateKey, publicKey, signingAlgorithm));
             assertEquals(200, response.getStatusCode());
 
-            // send a JWS using a algorithm other than the default (RS256)
+            // sending a JWS using another RSA based alg (ES256) should work as alg is not specified
```
Comment (tnorimat) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/ClientAuthSignedJWTTest.java:527]: "ES256" might be "PS256".

**Review by tnorimat — COMMENTED** [review]
@rmartinc I've added some review comments. Could you check them?

File: services/src/main/java/org/keycloak/crypto/ClientECDSASignatureVerifierContext.java [hunk 2]
```diff
@@ -16,6 +16,18 @@ private static KeyWrapper getKey(KeycloakSession session, ClientModel client, JW
...
+        if (!KeyType.EC.equals(key.getType())) {
+            throw new VerificationException("Key Type is not EC: " + key.getType());
+        }
+        if (key.getAlgorithm() == null) {
+            // defaults to the algorithm set to the JWS
+            // validations should be performed prior to verifying signature in case there are restrictions on the algorithms
+            // that can used for signing
+            key.setAlgorithm(input.getHeader().getRawAlgorithm());
+        } else if (!key.getAlgorithm().equals(input.getHeader().getRawAlgorithm())) {
+            throw new VerificationException("Key Algorithms are diferent, key-algorithm=" + key.getAlgorithm()
```
[the remark this answers is not in the document]
Comment (rmartinc) [reply services/src/main/java/org/keycloak/crypto/ClientECDSASignatureVerifierContext.java]: Done!

File: services/src/main/java/org/keycloak/crypto/ClientAsymmetricSignatureVerifierContext.java [hunk 3]
```diff
@@ -33,11 +33,17 @@ private static KeyWrapper getKey(KeycloakSession session, ClientModel client, JW
...
+        if (!KeyType.RSA.equals(key.getType())) {
+            throw new VerificationException("Key Type is not RSA: " + key.getType());
+        }
         if (key.getAlgorithm() == null) {
             // defaults to the algorithm set to the JWS
             // validations should be performed prior to verifying signature in case there are restrictions on the algorithms
             // that can used for signing
             key.setAlgorithm(input.getHeader().getRawAlgorithm());
+        } else if (!key.getAlgorithm().equals(input.getHeader().getRawAlgorithm())) {
+            throw new VerificationException("Key Algorithms are diferent, key-algorithm=" + key.getAlgorithm()
```
[the remark this answers is not in the document]
Comment (rmartinc) [reply services/src/main/java/org/keycloak/crypto/ClientAsymmetricSignatureVerifierContext.java]: Done!

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/ClientAuthSignedJWTTest.java (hunk 1 above)
Comment (rmartinc) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/ClientAuthSignedJWTTest.java]: Ups! Copy and paste! Modified!

Comment (tnorimat) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/ClientAuthSignedJWTTest.java]: Thanks!

File: services/src/main/java/org/keycloak/crypto/ClientAsymmetricSignatureVerifierContext.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (tnorimat) [reply services/src/main/java/org/keycloak/crypto/ClientAsymmetricSignatureVerifierContext.java]: Thanks!

File: services/src/main/java/org/keycloak/crypto/ClientECDSASignatureVerifierContext.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (tnorimat) [reply services/src/main/java/org/keycloak/crypto/ClientECDSASignatureVerifierContext.java]: Thanks!

**Review by tnorimat — DISMISSED** [review]

**Comment by tnorimat** [comment]
@mposolda Could you have a chance to check this PR?

**Comment by mposolda** [comment]
@rmartinc @tnorimat I am sorry for the late review of this PR.

PR looks quite good to me. Just one minor thing: There is new option `algorithm` added in the adapter credentials provider. It may be good to introduce it also to the `JWTClientAuthenticator.getAdapterConfiguration` and also possibly to the documentation https://www.keycloak.org/docs/latest/securing_apps/index.html#client-authentication-with-signed-jwt . The docs change will likely require small PR to the documentation. WDYT?

**Comment by tnorimat** [comment]
@mposolda Yes, I agree with your idea.

**Comment by rmartinc** [comment]
OK, @mposolda @tnorimat I have updated this PR with a few lines: [this](https://github.com/keycloak/keycloak/blob/62f4608837de84a9bc71fbac956dbd41a3ef1661/services/src/main/java/org/keycloak/authentication/authenticators/client/JWTClientAuthenticator.java#L262-L265) and [this](https://github.com/keycloak/keycloak/blob/62f4608837de84a9bc71fbac956dbd41a3ef1661/services/src/main/java/org/keycloak/authentication/authenticators/client/JWTClientSecretAuthenticator.java#L236-L239). And I also sent another PR for the documentation https://github.com/keycloak/keycloak-documentation/pull/1548 with a minor change.

There can be more thing more things to improve like generating also EC certificates in the endpoint (but that would need a new parameter for the type). You can always create the certificate yourself and upload the cert to the client as the PR is doing for tests.

Let me know if you see this is enough or you prefer to do more things.

**Review by tnorimat — APPROVED** [review]
LGTM

**Comment by mposolda** [comment]
@rmartinc @tnorimat Thanks for the update and the additional review!

**Pull request merged**